In [1]:
import sys

import pandas as pd
import numpy as np
from pathlib import Path
from matplotlib_venn import venn3

import matplotlib.pyplot as plt


sys.path.append(str(Path().resolve().parent))
from utils import create_venn_diagram

In [2]:
PATH = Path("../data/raw")

files = {
    "nodes": PATH / "nodes.csv",
    "streets": PATH / "streets.csv",
    "segments": PATH / "segments.csv",
    "status": PATH / "segment_status.csv",
    "train": PATH / "train.csv"
}

# Đọc dữ liệu

In [3]:
nodes_df = pd.read_csv(files["nodes"])
streets_df = pd.read_csv(files["streets"])
segments_df = pd.read_csv(files["segments"])
status_df = pd.read_csv(files["status"])
train_df = pd.read_csv(files["train"])

# Kiểu dữ liệu

In [4]:
nodes_df.dtypes

_id       int64
long    float64
lat     float64
dtype: object

In [5]:
streets_df.dtypes

_id               int64
level             int64
max_velocity    float64
name                str
type                str
dtype: object

In [6]:
segments_df.dtypes

_id               int64
created_at          str
updated_at          str
s_node_id         int64
e_node_id         int64
length            int64
street_id         int64
max_velocity    float64
street_level      int64
street_name         str
street_type         str
dtype: object

In [7]:
status_df.dtypes

_id           int64
updated_at      str
segment_id    int64
velocity      int64
dtype: object

In [8]:
train_df.dtypes

_id               int64
segment_id        int64
date                str
weekday           int64
period              str
LOS                 str
s_node_id         int64
e_node_id         int64
length            int64
street_id         int64
max_velocity    float64
street_level      int64
street_name         str
street_type         str
long_snode      float64
lat_snode       float64
long_enode      float64
lat_enode       float64
dtype: object

# Thống kê mô tả

In [9]:
nodes_df.describe()

,_id,long,lat
count,5.779670e+05,577967.000000,577967.000000
mean,4.315269e+09,106.703700,10.808332
std,1.887984e+09,0.069665,0.049736
min,3.663672e+08,106.536005,10.676001
25%,2.964647e+09,106.655452,10.772766
50%,5.627164e+09,106.714413,10.824470
75%,5.737770e+09,106.747359,10.845259
max,6.205236e+09,106.867994,10.888998


In [10]:
streets_df.describe()

,_id,level,max_velocity
count,5.553000e+03,5553.000000,987.000000
mean,3.457455e+08,3.199712,58.232016
std,2.017849e+08,0.991818,16.046644
min,3.109679e+07,1.000000,10.000000
25%,1.991616e+08,3.000000,50.000000
50%,3.215683e+08,4.000000,60.000000
75%,5.682871e+08,4.000000,60.000000
max,6.583281e+08,4.000000,120.000000


In [11]:
segments_df.describe()

,_id,s_node_id,e_node_id,length,street_id,max_velocity,street_level
count,84633.000000,8.463300e+04,8.463300e+04,84633.000000,8.463300e+04,9871.000000,84633.000000
mean,42316.000000,3.674041e+09,3.672725e+09,41.818050,2.866904e+08,54.503596,3.500006
std,24431.587003,2.039756e+09,2.040056e+09,52.692336,2.085526e+08,16.553367,0.841222
min,0.000000,3.663672e+08,3.663672e+08,0.000000,3.109679e+07,10.000000,1.000000
25%,21158.000000,2.078951e+09,2.078951e+09,13.000000,3.511357e+07,40.000000,3.000000
50%,42316.000000,4.068031e+09,4.068031e+09,27.000000,2.517313e+08,60.000000,4.000000
75%,63474.000000,5.748273e+09,5.748270e+09,52.000000,4.693276e+08,60.000000,4.000000
max,84632.000000,6.175932e+09,6.175932e+09,1622.000000,6.583281e+08,120.000000,4.000000


In [12]:
status_df.describe()

,_id,segment_id,velocity
count,90938.000000,90938.000000,90938.000000
mean,45468.500000,40376.030009,20.754844
std,26251.683727,22415.108133,19.567112
min,0.000000,26.000000,0.000000
25%,22734.250000,21263.000000,2.000000
50%,45468.500000,41875.000000,17.000000
75%,68202.750000,58440.750000,34.000000
max,90937.000000,84535.000000,200.000000


In [13]:
train_df.describe()

,_id,segment_id,weekday,s_node_id,e_node_id,length,street_id,max_velocity,street_level,long_snode,lat_snode,long_enode,lat_enode
count,33441.000000,33441.000000,33441.000000,3.344100e+04,3.344100e+04,33441.000000,3.344100e+04,4946.000000,33441.000000,33441.000000,33441.000000,33441.000000,33441.000000
mean,16720.000000,41014.936575,2.915254,3.528928e+09,3.582345e+09,48.709010,2.715539e+08,59.299434,3.029993,106.670354,10.804326,106.670334,10.804352
std,9653.729512,22489.675329,2.329021,2.008085e+09,2.014716e+09,52.251062,1.908930e+08,14.172472,1.022137,0.041887,0.040358,0.041868,0.040377
min,0.000000,26.000000,0.000000,3.663674e+08,3.663678e+08,0.000000,3.257582e+07,10.000000,1.000000,106.537476,10.676452,106.537776,10.676467
25%,8360.000000,21967.000000,1.000000,2.079964e+09,2.079965e+09,17.000000,5.588046e+07,50.000000,2.000000,106.649521,10.772135,106.649478,10.772131
50%,16720.000000,41654.000000,3.000000,3.771416e+09,4.066314e+09,34.000000,2.420036e+08,60.000000,3.000000,106.664687,10.797884,106.664704,10.797776
75%,25080.000000,58456.000000,5.000000,5.738236e+09,5.748445e+09,63.000000,4.082464e+08,60.000000,4.000000,106.685096,10.837117,106.685212,10.837388
max,33440.000000,84535.000000,6.000000,6.175932e+09,6.175932e+09,1011.000000,6.565625e+08,120.000000,4.000000,106.867740,10.888965,106.867654,10.888993


# Kiểm tra dữ liệu thừa thiếu

## nodes.csv

In [14]:
print(nodes_df.duplicated().sum())
nodes_df.isnull().sum()

0


_id     0
long    0
lat     0
dtype: int64

## streets.csv

In [15]:
print(streets_df.duplicated().sum())
streets_df.isnull().sum()

0


_id                0
level              0
max_velocity    4566
name            1605
type               0
dtype: int64

In [16]:
streets_df.isnull().sum() * 100 / len(streets_df)

_id              0.000000
level            0.000000
max_velocity    82.225824
name            28.903296
type             0.000000
dtype: float64

In [17]:
# Thế max_velocity = -1 cho giá trị bị thiếu
streets_df["max_velocity"] = streets_df["max_velocity"].fillna(40)

In [18]:
# Thế name = "None" cho giá trị bị thiếu
streets_df["name"] = streets_df["name"].fillna("None")

In [19]:
streets_df.isnull().sum()

_id             0
level           0
max_velocity    0
name            0
type            0
dtype: int64

## segments.csv

In [20]:
print(segments_df.duplicated().sum())
segments_df.isnull().sum()

0


_id                 0
created_at          0
updated_at          0
s_node_id           0
e_node_id           0
length              0
street_id           0
max_velocity    74762
street_level        0
street_name       152
street_type         0
dtype: int64

In [21]:
segments_df.isnull().sum() * 100 / len(segments_df)

_id              0.000000
created_at       0.000000
updated_at       0.000000
s_node_id        0.000000
e_node_id        0.000000
length           0.000000
street_id        0.000000
max_velocity    88.336701
street_level     0.000000
street_name      0.179599
street_type      0.000000
dtype: float64

In [22]:
segments_df["max_velocity"] = segments_df["max_velocity"].fillna(40)

In [23]:
segments_df["street_name"] = segments_df["street_name"].fillna("None")

In [24]:
segments_df.isnull().sum()

_id             0
created_at      0
updated_at      0
s_node_id       0
e_node_id       0
length          0
street_id       0
max_velocity    0
street_level    0
street_name     0
street_type     0
dtype: int64

## segment_status.csv

In [25]:
print(status_df.duplicated().sum())
status_df.isnull().sum()

0


_id           0
updated_at    0
segment_id    0
velocity      0
dtype: int64

## train.csv

In [26]:
print(train_df.duplicated().sum())
train_df.isnull().sum()

0


_id                 0
segment_id          0
date                0
weekday             0
period              0
LOS                 0
s_node_id           0
e_node_id           0
length              0
street_id           0
max_velocity    28495
street_level        0
street_name         1
street_type         0
long_snode          0
lat_snode           0
long_enode          0
lat_enode           0
dtype: int64

In [27]:
train_df.isnull().sum() * 100 / len(train_df)

_id              0.000000
segment_id       0.000000
date             0.000000
weekday          0.000000
period           0.000000
LOS              0.000000
s_node_id        0.000000
e_node_id        0.000000
length           0.000000
street_id        0.000000
max_velocity    85.209772
street_level     0.000000
street_name      0.002990
street_type      0.000000
long_snode       0.000000
lat_snode        0.000000
long_enode       0.000000
lat_enode        0.000000
dtype: float64

In [28]:
train_df["max_velocity"] = train_df["max_velocity"].fillna(40)

In [29]:
train_df["street_name"] = train_df["street_name"].fillna("None")

In [30]:
train_df.isnull().sum()

_id             0
segment_id      0
date            0
weekday         0
period          0
LOS             0
s_node_id       0
e_node_id       0
length          0
street_id       0
max_velocity    0
street_level    0
street_name     0
street_type     0
long_snode      0
lat_snode       0
long_enode      0
lat_enode       0
dtype: int64

In [31]:
print(
    nodes_df.shape[1] +
    segments_df.shape[1] +
    status_df.shape[1] +
    streets_df.shape[1] +
    train_df.shape[1]
)


41


# Kiểm tra tính nhất quán của dữ liệu

## Điều kiện

$|\pi_{\_id}(nodes)| = |nodes|$<br>
$|\pi_{\_id}(segments)| = |segments|$<br>
$|\pi_{\_id}(segments\_status)| = |segments\_status|$<br>
$|\pi_{\_id}(streets)| = |streets|$<br>
$|\pi_{\_id}(train)| = |train|$

$\pi_{s\_node\_id}(segments) \subseteq \pi_{\_id}(nodes)$<br>
$\pi_{e\_node\_id}(segments) \subseteq \pi_{\_id}(nodes)$

$\pi_{s\_node\_id}(train) \subseteq \pi_{s\_node\_id}(segments)$<br>
$\pi_{e\_node\_id}(train) \subseteq \pi_{e\_node\_id}(segments)$

$\pi_{s\_node\_id, long\_snode, lat\_snode}(train) \subseteq nodes$<br>
$\pi_{e\_node\_id, long\_enode, lat\_enode}(train) \subseteq nodes$

$\pi_{segment\_id}(segment\_status) \subseteq \pi_{\_id}(segments)$

$\pi_{segment\_id, length}(train) \subseteq \pi_{\_id, length}(segments)$

$P = {\pi_{street_id, street_level, max_velocity, street_name, street_type}(train)}$

$SegTrain \subseteq SegSegments$
$SegTrain = \pi_{street\_id, street\_level, max\_velocity, street\_name, street\_type}(train)$
$SegSegments = \pi_{street\_id, street\_level, max\_velocity, street\_name, street\_type}(segments)$

$StSegments \subseteq streets$<br>
$StTrain \subseteq streets$<br>
$StSegments = \pi_{street\_id, street\_level, max\_velocity, street\_name, street\_type}(segments)$<br>
$StTrain = \pi_{street\_id, street\_level, max\_velocity, street\_name, street\_type}(train)$

## Đảm bảo id là duy nhất

### Khóa chính của nodes

In [32]:
print("Số lượng id không bị trùng:", nodes_df["_id"].duplicated().sum())
nodes = set(nodes_df["_id"]) # Tập các nút
n_nodes = len(nodes)
print("Số lượng nút có trong nodes.csv:", n_nodes)

Số lượng id không bị trùng: 0
Số lượng nút có trong nodes.csv: 577967


### Khóa chính của segments

In [33]:
print("Số lượng _id bị trùng:", segments_df["_id"].duplicated().sum())
segments = set(segments_df["_id"])
n_segments = len(segments)
print("Số lượng segments trong segments.csv:", n_segments)

Số lượng _id bị trùng: 0
Số lượng segments trong segments.csv: 84633


### Khóa chính của segments_status

In [34]:
print("Số lượng _id bị trùng:", status_df["_id"].duplicated().sum())
status_id = set(status_df["_id"])
n_status = len(status_df)
print("Số lượng status id trong segments_status:", n_status)

Số lượng _id bị trùng: 0
Số lượng status id trong segments_status: 90938


### Khóa chính của streets

In [35]:
print("Số lượng _id bị trùng:", streets_df["_id"].duplicated().sum())
streets = set(streets_df["_id"])
n_streets = len(streets)
print("Số lượng street trong streets.csv:", n_streets)

Số lượng _id bị trùng: 0
Số lượng street trong streets.csv: 5553


### Khóa chính của train

In [36]:
print("Số lượng _id bị trùng:", train_df["_id"].duplicated().sum())
train_id = set(train_df["_id"])
n_train = len(train_id)
print("Số lượng street trong streets.csv:", n_train)

Số lượng _id bị trùng: 0
Số lượng street trong streets.csv: 33441


## Tính nhất quản của node

### Tập s_node_id, e_node_id trong segments và nodes

In [37]:
segments_snodes = set(segments_df["s_node_id"])
n_s_segments_nodes = len(segments_snodes)

inter_s_nodes_segments = nodes & segments_snodes
print("Số lượng nút trong nodes:", n_nodes)
print("Số lượng nút nguồn của segments:", len(segments_snodes))
print("Số lượng s_node giao với node:", len(inter_s_nodes_segments))

Số lượng nút trong nodes: 577967
Số lượng nút nguồn của segments: 52384
Số lượng s_node giao với node: 52384


$|\pi_{\_id}(nodes)| = 577967$<br>
$|\pi_{s\_node\_id}(segments)| = 52384$ <br>
Do $|\pi_{\_id}(nodes) \cap \pi_{s\_ node\_id}(segments)| = |\pi_{s\_node\_id}(segments)| = 52384$<br>
$\Rightarrow \pi_{\_id}(segments) \subset \pi_{s\_node\_id}(nodes)$

Do $|\pi_{\_id}(nodes) \cap \pi_{s\_node\_id}(segments)| = \pi_{s\_node\_id}(segments)$<br>
$\Rightarrow \pi_{\_id}(segments) \subset \pi_{s\_node\_id}(nodes)$

In [38]:
segments_enodes = set(segments_df["e_node_id"])

inter_e_nodes_segments = nodes & segments_enodes
print("Số lượng nút trong nodes:", n_nodes)
print("Số lượng nút đích của segments:", len(segments_enodes))
print("Số lượng e_node giao với node:", len(inter_e_nodes_segments))

Số lượng nút trong nodes: 577967
Số lượng nút đích của segments: 52386
Số lượng e_node giao với node: 52386


$|\pi_{\_id}(nodes)| = 577967$<br>
$|\pi_{e\_node\_id}(segments)| = 52384$ <br>
Do $|\pi_{\_id}(nodes) \cap \pi_{s\_node\_id}(segments)| = |\pi_{e\_node\_id}(segments)| = 52384$<br>
$\Rightarrow \pi_{\_id}(segments) \subset \pi_{e\_node\_id}(nodes)$

Do $\pi_{\_id}(nodes) \cap \pi_{e\_node\_id}(segments) = \pi_{e\_node\_id}(segments)$<br>
$\Rightarrow \pi_{\_id}(segments) \subset \pi_{e\_node\_id}(nodes)$

In [39]:
###

### Tập s_node_id, e_node_id trong train và segments

In [40]:
train_snodes = set(train_df["s_node_id"])
train_enodes = set(train_df["e_node_id"])

In [41]:
inter_snode_train_segments = train_snodes & segments_snodes
print("Số lượng nút nguồn trong segments:", len(segments_snodes))
print("Số lượng nút nguồn của train:", len(train_snodes))
print("Số lượng e_node giao với node:", len(inter_snode_train_segments))

Số lượng nút nguồn trong segments: 52384
Số lượng nút nguồn của train: 8430
Số lượng e_node giao với node: 8430


In [42]:
inter_enode_train_segments = train_enodes & segments_enodes
print("Số lượng nút nguồn trong segments:", len(segments_enodes))
print("Số lượng nút nguồn của train:", len(train_enodes))
print("Số lượng e_node giao với node:", len(inter_enode_train_segments))

Số lượng nút nguồn trong segments: 52386
Số lượng nút nguồn của train: 8427
Số lượng e_node giao với node: 8427


### Tập s_node_id, e_node_id trong train và nodes

In [43]:
inter_snode_train_nodes = train_snodes & nodes
print("Số lượng nút:", len(nodes))
print("Số lượng nút nguồn của train:", len(train_snodes))
print("Số lượng e_node giao với node:", len(inter_snode_train_nodes))

Số lượng nút: 577967
Số lượng nút nguồn của train: 8430
Số lượng e_node giao với node: 8430


In [44]:
inter_enode_train_nodes = train_enodes & nodes
print("Số lượng nút:", len(nodes))
print("Số lượng nút nguồn của train:", len(train_enodes))
print("Số lượng e_node giao với node:", len(inter_enode_train_nodes))

Số lượng nút: 577967
Số lượng nút nguồn của train: 8427
Số lượng e_node giao với node: 8427


### Tập s_node_id, long_snode, lat_snode

In [45]:
nodes_set = set(nodes_df.apply(tuple, axis=1))

In [46]:
train_snode_loc = set(train_df[["s_node_id", "long_snode", "lat_snode"]].apply(tuple, axis=1))
inter_train_snode_loc = nodes_set & train_snode_loc
print("Số node của nodes.csv:", len(nodes_set))
print("Số phần tử tập snode loc của train:", len(train_snode_loc))
print("Số phần tử tập giao snode loc của train và nodes:", len(inter_train_snode_loc))

Số node của nodes.csv: 577967
Số phần tử tập snode loc của train: 8430
Số phần tử tập giao snode loc của train và nodes: 8430


### Tập e_node_id, long_snode, lat_snode

In [47]:
train_enode_loc = set(train_df[["e_node_id", "long_enode", "lat_enode"]].apply(tuple, axis=1))
inter_enode_loc_train_nodes = nodes_set & train_enode_loc
print("Số node của nodes.csv:", len(nodes_set))
print("Số phần tử tập enode loc của train:", len(train_enode_loc))
print("Số phần tử tập giao enode loc của train và nodes:", len(inter_enode_loc_train_nodes))

Số node của nodes.csv: 577967
Số phần tử tập enode loc của train: 8427
Số phần tử tập giao enode loc của train và nodes: 8427


## Tính nhât quán của segments

### Tập node_id của segments và segment_status

In [48]:
segment_status = set(status_df["segment_id"])
inter_segment_status_segments = segment_status & segments
print("Số segment:", n_segments)
print("Số phần tử tập segment_id trong segment_status:", len(segment_status))
print("Số phần tử tập giao:", len(inter_segment_status_segments))

Số segment: 84633
Số phần tử tập segment_id trong segment_status: 10027
Số phần tử tập giao: 10027


### Tập các thuộc tính của segments và train

In [49]:
train_segment = set(train_df[
        ["segment_id", 
        "s_node_id", 
        "e_node_id", 
        "length", 
        "street_id", 
        "max_velocity", 
        "street_level", 
        "street_name"]
    ].apply(tuple, axis=1)
)

segments_segment = set(segments_df[
        ["_id", 
        "s_node_id", 
        "e_node_id", 
        "length", 
        "street_id", 
        "max_velocity", 
        "street_level", 
        "street_name"]
    ].apply(tuple, axis=1)
)

inter_segments_train_segment = train_segment & segments_segment
print("Số segment trong segments:", len(segments_segment))
print("Số segment trong train:", len(train_segment))
print("Số segment trong tập giao:", len(inter_segments_train_segment))

Số segment trong segments: 84633
Số segment trong train: 10027
Số segment trong tập giao: 10027


## Tính nhất quán của street

In [50]:
streets_street = set(
    streets_df[[
        "_id",
        "level",
        "max_velocity",
        "name",
        "type"
    ]].apply(tuple, axis=1)
)

segments_street = set(
    segments_df[[
        "street_id",
        "street_level",
        "max_velocity",
        "street_name",
        "street_type"
    ]].apply(tuple, axis=1)
)

In [51]:
inter_segments_streets_street = streets_street & segments_street
print("Số street trong streets", len(streets_street))
print("Số street trong segments", len(segments_street))
print("Số street trong tập giao", len(inter_segments_streets_street))

Số street trong streets 5553
Số street trong segments 6301
Số street trong tập giao 3954


In [52]:
train_street = set(
    train_df[[
        "street_id",
        "street_level",
        "max_velocity",
        "street_name",
        "street_type"
    ]].apply(tuple, axis=1)
)

In [53]:
inter_segments_train_street = segments_street & train_street
print("Số street trong segments", len(segments_street))
print("Số street trong train", len(train_street))
print("Số street trong tập giao", len(inter_segments_train_street))

Số street trong segments 6301
Số street trong train 2031
Số street trong tập giao 2031


#### Kiểm tra tổng quan

In [54]:
temp_streets_df = streets_df[["_id", "name", "level", "max_velocity", "type"]].copy()
temp_segments_df = segments_df[["street_id", "street_name", "street_level", "max_velocity", "street_type"]].copy()
temp_segments_df = temp_segments_df.rename(
    columns={
        "street_id": "_id",
        "street_name": "name",
        "street_level": "level",
        "street_type": "type"
})

In [55]:
conflicts_df = pd.concat([temp_streets_df, temp_segments_df]).drop_duplicates()
print(conflicts_df.shape)
group_id_conflicts = conflicts_df.groupby("_id")

(7900, 5)


In [56]:
# In vài dữ liệu dòng dữ liệu
count = 0
for _id, group in group_id_conflicts:
    if count == 9:
        break
        
    if len(group) > 1:
        print(f"\n=== ID: {_id}  |  Conflicts: {len(group)} ===")
        print(group.to_string(index=False))
        print("-" * 80)
        count += 1


=== ID: 32575737  |  Conflicts: 3 ===
     _id           name  level  max_velocity         type
32575737           None      4          40.0 unclassified
32575737     Thương Mại      4          40.0  marketplace
32575737 Xã Trung Chánh      4          40.0 unclassified
--------------------------------------------------------------------------------

=== ID: 32575874  |  Conflicts: 2 ===
     _id           name  level  max_velocity         type
32575874           None      4          40.0 unclassified
32575874 Xã Trung Chánh      4          40.0 unclassified
--------------------------------------------------------------------------------

=== ID: 32576095  |  Conflicts: 3 ===
     _id               name  level  max_velocity     type
32576095               None      4          40.0 tertiary
32576095 đường Thái Thị Giữ      4          40.0 tertiary
32576095         Xã Bà Điểm      4          40.0 tertiary
--------------------------------------------------------------------------------

=

In [57]:
# In các dòng mâu thuẫn nhiều hơn 2
count = 0
for _id, group in group_id_conflicts:
    if count == 9:
        break
        
    if len(group) > 2:
        print(f"\n=== ID: {_id}  |  Conflicts: {len(group)} ===")
        print(group.to_string(index=False))
        print("-" * 80)
        count += 1


=== ID: 32575737  |  Conflicts: 3 ===
     _id           name  level  max_velocity         type
32575737           None      4          40.0 unclassified
32575737     Thương Mại      4          40.0  marketplace
32575737 Xã Trung Chánh      4          40.0 unclassified
--------------------------------------------------------------------------------

=== ID: 32576095  |  Conflicts: 3 ===
     _id               name  level  max_velocity     type
32576095               None      4          40.0 tertiary
32576095 đường Thái Thị Giữ      4          40.0 tertiary
32576095         Xã Bà Điểm      4          40.0 tertiary
--------------------------------------------------------------------------------

=== ID: 32577095  |  Conflicts: 3 ===
     _id             name  level  max_velocity     type
32577095             None      4          40.0 tertiary
32577095         Trần Não      4          40.0 tertiary
32577095 Phường Thảo Điền      4          40.0 tertiary
---------------------------------

In [58]:
# Tìm một trong dòng trong đó con đường không được đặt tên
count = 0
for _id, group in conflicts_df.groupby('_id'):
    if count == 4:
        break
    
    if len(group) > 1:
        if (group["name"] != "None").all():
            print(f"\n=== ID: {_id} | Conflicts: {len(group)} ===")
            print(group[["_id", "name"]].to_string(index=False))
            print("-" * 80)
            count += 1

Như vậy trong tất cả các trường hợp đều có đường không được đặt tên. Từ kiểm tra tổng quát có thể gặp 2 trường hợp với 2 thuộc tính street_name và street_type:
- Đối với street_name: Cùng một id nhưng lại có tên đường khác nhau
- Đối với street_type: Cùng một id nhưng lại có loại đường khác nhau
Phương pháp giải quyết là tạo ra một streets mới

#### Kiểm tra từng thuộc tính

In [59]:
# Thuộc tính street_level
street_level_streets = set(streets_df[["_id", "level"]].apply(tuple, axis=1))
street_level_segments = set(segments_df[["street_id", "street_level"]].apply(tuple, axis=1))

inter_street_level_streets_segments = (
    street_level_streets & 
    street_level_segments
)

print("Số phần tử của tập street_level trong segments.csv:", len(street_level_streets))
print("Số phần tử của tập street_level của streets.csv:", len(street_level_segments))
print("Số phần tử của tập giao:", len(inter_street_level_streets_segments))

Số phần tử của tập street_level trong segments.csv: 5553
Số phần tử của tập street_level của streets.csv: 5539
Số phần tử của tập giao: 5539


$\begin{align*} 
\pi_{\_id, level}(streets) \cap \pi_{street\_id, street\_level}(segments) \\ = \pi_{street\_id, street\_level}(segments) 
\end{align*}$

In [60]:
max_velocity_streets = set(streets_df[["_id", "max_velocity"]].apply(tuple, axis=1))
max_velocity_segments = set(segments_df[["street_id", "max_velocity"]].apply(tuple, axis=1))

inter_max_velocity_streets_segments = (
    max_velocity_streets & 
    max_velocity_segments
)

print("Số phần tử của tập max_velocity của streets.csv:", len(max_velocity_streets))
print("Số phần tử của tập max_velocity của segments.csv:", len(max_velocity_segments))
print("Số phần tử của tập giao:", len(inter_max_velocity_streets_segments))

Số phần tử của tập max_velocity của streets.csv: 5553
Số phần tử của tập max_velocity của segments.csv: 5539
Số phần tử của tập giao: 5539


$\begin{align*}
\pi_{\_id, max\_velocity}(streets) \cap \pi_{street\_id, max\_velocity}(segments) \\
= \pi_{street\_id, max\_velocity}(segments)
\end{align*}$

In [61]:
street_type_streets = set(streets_df[["_id", "type"]].apply(tuple, axis=1))
street_type_segments = set(segments_df[["street_id", "street_type"]].apply(tuple, axis=1))

inter_street_type_streets_segments = (
    street_type_streets & 
    street_type_segments
)
print("Số phần tử của tập streets của streets.csv:", len(street_type_streets))
print("Số phần tử của tập streets của segments.csv:", len(street_type_segments))
print("Số phần tử của tập giao:", len(inter_street_type_streets_segments))

Số phần tử của tập streets của streets.csv: 5553
Số phần tử của tập streets của segments.csv: 5816
Số phần tử của tập giao: 5422


$\begin{align*}
|\pi_{\_id, type}(streets)| = 5553 \\
|\pi_{\_id, street\_type}(segments)| = 5816 \\
|\pi_{\_id, type}(streets)| \cap |\pi_{\_id, street\_type}(segments)| = 5422
\end{align*}$

In [62]:
street_name_streets = set(streets_df[["_id", "name"]].apply(tuple, axis=1))
street_name_segments = set(segments_df[["street_id", "street_name"]].apply(tuple, axis=1))

inter_street_type_streets_segments = (
    street_name_streets & 
    street_name_segments
)
print("Số phần tử của tập streets của streets.csv:", len(street_name_streets))
print("Số phần tử của tập streets của segments.csv:", len(street_name_segments))
print("Số phần tử của tập giao:", len(inter_street_type_streets_segments))

Số phần tử của tập streets của streets.csv: 5553
Số phần tử của tập streets của segments.csv: 6103
Số phần tử của tập giao: 3954


$\begin{align*}
|\pi_{\_id, name}(streets)| = 5553 \\
|\pi_{\_id, street\_name}(segments)| = 6103 \\
|\pi_{\_id, name}(streets)| \cap |\pi_{\_id, street\_name}(segments)| = 3954
\end{align*}$

#### Phân tích các thuộc tính bị xung đột

##### Thuộc tính type

In [63]:
unique_segments_type = segments_df["street_type"].unique()
unique_streets_type = streets_df["type"].unique()

print("Số unique các loại đường trong segments.csv:", len(unique_streets_type))
print("Số unique các loại đường trong streets.csv:", len(unique_segments_type))

Số unique các loại đường trong segments.csv: 11
Số unique các loại đường trong streets.csv: 53


In [64]:
print("Các loại đường của segments:")
print(np.sort(unique_segments_type))
print("Các loại đường của streets:")
print(np.sort(unique_streets_type))

Các loại đường của segments:
['apartments' 'atm' 'bank' 'bridge' 'bus_station' 'bus_stop' 'car'
 'chalet' 'church' 'cinema' 'clothes' 'college' 'community_centre'
 'company' 'construction' 'convenience' 'dentist' 'department_store'
 'dormitory' 'dry_cleaning' 'fast_food' 'fire_station' 'fuel' 'government'
 'hospital' 'hostel' 'house' 'industrial' 'marketplace' 'motorway'
 'motorway_link' 'pedestrian' 'pitch' 'police' 'primary' 'primary_link'
 'residential' 'restaurant' 'school' 'secondary' 'secondary_link'
 'service' 'shelter' 'swimming_pool' 'tertiary' 'tertiary_link'
 'theme_park' 'training_area' 'trunk' 'trunk_link' 'unclassified'
 'university' 'yes']
Các loại đường của streets:
['motorway' 'motorway_link' 'primary' 'primary_link' 'secondary'
 'secondary_link' 'tertiary' 'tertiary_link' 'trunk' 'trunk_link'
 'unclassified']


Như vậy có thể street_type trong segments là thuộc tính của segments, không phải của street

Qua đó xác nhận rằng thuộc tính street_type trong segments là thuộc tính của segments

In [65]:
count = 0
street_type_segments_conflicts_df = segments_df[["_id", "street_id", "street_type"]].drop_duplicates()
street_type_segments_conflicts_group = street_type_segments_conflicts_df.groupby("street_id")

for _id, group in street_type_segments_conflicts_group:
    # Kiểm tra nếu tất cả street_type trong group đều giống nhau thì bỏ qua
    if group["street_type"].nunique() == 1:
        continue   # Bỏ qua group không có conflict
    
    if count == 4:
        break
    
    print(f"\n=== ID: {_id} | Conflicts: {len(group)} ===")
    print(group.to_string(index=False))
    print("-" * 50)
    count += 1


=== ID: 32575737 | Conflicts: 2 ===
 _id  street_id  street_type
  10   32575737  marketplace
  11   32575737 unclassified
--------------------------------------------------

=== ID: 32578248 | Conflicts: 22 ===
 _id  street_id  street_type
3369   32578248  bus_station
3370   32578248         bank
3371   32578248 unclassified
3372   32578248 unclassified
3373   32578248 unclassified
3374   32578248 unclassified
3375   32578248 unclassified
3376   32578248 unclassified
3377   32578248 unclassified
3378   32578248 unclassified
3379   32578248 unclassified
3380   32578248 unclassified
3381   32578248 unclassified
3382   32578248 unclassified
3383   32578248 unclassified
3384   32578248 unclassified
3385   32578248 unclassified
3386   32578248 unclassified
3387   32578248 unclassified
3388   32578248 unclassified
3389   32578248 unclassified
3390   32578248 unclassified
--------------------------------------------------

=== ID: 32578371 | Conflicts: 82 ===
 _id  street_id street_type
340

##### Thuộc tính name

In [66]:
unique_segments_name = set(segments_df["street_name"].unique())
unique_streets_name = set(streets_df["name"].unique())

print("Số unique các loại đường trong segments.csv:", len(unique_segments_name))
print("Số unique các loại đường trong streets.csv:", len(unique_streets_name))

Số unique các loại đường trong segments.csv: 1389
Số unique các loại đường trong streets.csv: 1049


In [67]:
count = 0
street_name_segments_conflicts_df = segments_df[["_id", "street_id", "street_name"]].drop_duplicates()
street_name_segments_conflicts_group = street_name_segments_conflicts_df.groupby("street_id")

for _id, group in street_name_segments_conflicts_group:
    # Kiểm tra nếu tất cả street_type trong group đều giống nhau thì bỏ qua
    if group["street_name"].nunique() == 1:
        continue   # Bỏ qua group không có conflict
    
    if count == 4:
        break

    group = group.drop_duplicates()
    print(f"\n=== ID: {_id} | Conflicts: {len(group)} ===")
    print(group.to_string(index=False))
    print("-" * 50)
    count += 1


=== ID: 32575737 | Conflicts: 2 ===
 _id  street_id    street_name
  10   32575737     Thương Mại
  11   32575737 Xã Trung Chánh
--------------------------------------------------

=== ID: 32576095 | Conflicts: 36 ===
 _id  street_id        street_name
 331   32576095 đường Thái Thị Giữ
 332   32576095 đường Thái Thị Giữ
 333   32576095 đường Thái Thị Giữ
 334   32576095         Xã Bà Điểm
 335   32576095 đường Thái Thị Giữ
 336   32576095 đường Thái Thị Giữ
 337   32576095 đường Thái Thị Giữ
 338   32576095 đường Thái Thị Giữ
 339   32576095 đường Thái Thị Giữ
 340   32576095 đường Thái Thị Giữ
 341   32576095 đường Thái Thị Giữ
 342   32576095 đường Thái Thị Giữ
 343   32576095 đường Thái Thị Giữ
 344   32576095 đường Thái Thị Giữ
 345   32576095 đường Thái Thị Giữ
 346   32576095 đường Thái Thị Giữ
 347   32576095 đường Thái Thị Giữ
 348   32576095 đường Thái Thị Giữ
 349   32576095 đường Thái Thị Giữ
 350   32576095 đường Thái Thị Giữ
 351   32576095 đường Thái Thị Giữ
 352   3257

Trong số các tên quan sát được, thì chỉ có một tên đường bị sai dữ liệu.

#### Giải pháp 

Thêm id mới cho streets

In [68]:
new_streets_df = streets_df.copy()
new_train_df = train_df.copy()
new_segments_df = segments_df.copy()

In [69]:
# Lấy index của cột "name"
streets_name_idx = new_streets_df.columns.get_loc("name")

# Lấy index cho cột "name", "type"
segments_id_idx = new_segments_df.columns.get_loc("_id")

streets_types = new_streets_df["type"].unique()

for _id, group in conflicts_df.groupby('_id'):
    if len(group) >= 2:
        # Đổi tên new_streets_df.name theo hàng 2
        # Lấy tên đường mới
        street_name = group["name"].to_list()[1]
        
        # Lấy index của row cần sửa trong new_streets_df
        row_idx = new_streets_df[new_streets_df["_id"] == _id].index[0]
        
        # Cập nhật giá trị
        new_streets_df.iloc[row_idx, streets_name_idx] = street_name

        # Thêm id cho các hàng còn lại nếu có
        if len(group) > 2:
            rows = group.iloc[2:,:]
            for row in rows.itertuples():
                # Thêm hàng cho streets.csv
                new_id = new_streets_df["_id"].max() + 1
                
                new_row = pd.Series({
                    "_id": new_id,
                    "name": row.name,
                    "level": row.level,
                    "type": row.type if row.type in streets_types else "unclassified",
                    "max_velocity": row.max_velocity
                })
                
                # Thêm vào test_df
                new_streets_df = pd.concat([new_streets_df, new_row.to_frame().T], ignore_index=True)

                # Sửa lại cho segments.csv
                segments_mask = (
                    (new_segments_df["street_id"] == _id) &
                    (new_segments_df["street_name"] == row.name) &
                    (new_segments_df["street_type"] == row.type)
                )
                new_segments_df.loc[segments_mask, "street_id"] = new_id

                # Sửa cho train.csv
                train_mask = (
                    (new_train_df["street_id"] == _id) &
                    (new_train_df["street_name"] == row.name) &
                    (new_train_df["street_type"] == row.type)
                )
                new_train_df.loc[train_mask, "street_id"] = new_id

##### Đổi tên segments.street_type -> segments.type

In [70]:
new_segments_df = new_segments_df.rename(columns={"street_type": "type"})

In [71]:
new_streets_df.columns

Index(['_id', 'level', 'max_velocity', 'name', 'type'], dtype='str')

##### Đổi tên train.street_type -> train.segment_type

In [72]:
new_train_df = new_train_df.rename(columns={"street_type": "segment_type"})
new_train_df.head()

,_id,segment_id,date,weekday,period,LOS,s_node_id,e_node_id,length,street_id,max_velocity,street_level,street_name,segment_type,long_snode,lat_snode,long_enode,lat_enode
0,0,26,2021-04-16,4,period_0_30,A,366428456,366416066,116,32575820,40.0,4,Nguyễn Văn Bá,tertiary,106.768732,10.841506,106.769254,10.842422
1,1,33,2020-08-02,6,period_23_30,C,366469460,3792257828,26,32575862,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808
2,2,33,2020-08-03,0,period_0_00,D,366469460,3792257828,26,32575862,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808
3,3,67,2021-03-09,1,period_9_30,B,366403668,5755066033,7,32575862,40.0,3,Đường số 5,secondary,106.768412,10.880817,106.768461,10.880771
4,4,67,2021-03-23,1,period_9_30,B,366403668,5755066033,7,32575862,40.0,3,Đường số 5,secondary,106.768412,10.880817,106.768461,10.880771


## Xác nhận lại

In [73]:
streets_street = set(
    new_streets_df[[
        "_id",
        "level",
        "max_velocity",
        "name"
    ]].apply(tuple, axis=1)
)

segments_street = set(
    new_segments_df[[
        "street_id",
        "street_level",
        "max_velocity",
        "street_name"
    ]].apply(tuple, axis=1)
)

inter_segments_streets_street = streets_street & segments_street
print("Số street trong streets", len(streets_street))
print("Số street trong segments", len(segments_street))
print("Số street trong tập giao", len(inter_segments_streets_street))

Số street trong streets 6312
Số street trong segments 6298
Số street trong tập giao 6298


## Xóa các thuộc tính không cần thiết

In [74]:
segments_df.head(1)

,_id,created_at,updated_at,s_node_id,e_node_id,length,street_id,max_velocity,street_level,street_name,street_type
0,0,2020-10-18T13:26:17.365Z,2020-10-18T13:26:17.365Z,373543511,5468660805,114,31096786,80.0,1,Quốc Lộ 1,trunk


In [75]:
filtered_segments_df = new_segments_df.drop(columns=["max_velocity", "street_level", "street_name"])
new_segments_df.head()

,_id,created_at,updated_at,s_node_id,e_node_id,length,street_id,max_velocity,street_level,street_name,type
0,0,2020-10-18T13:26:17.365Z,2020-10-18T13:26:17.365Z,373543511,5468660805,114,31096786,80.0,1,Quốc Lộ 1,trunk
1,1,2020-10-18T13:26:17.400Z,2020-10-18T13:26:17.400Z,5468660805,5738158916,9,31096786,80.0,1,Quốc Lộ 1,trunk
2,2,2020-10-18T13:26:17.435Z,2020-10-18T13:26:17.435Z,5738158916,5738158918,23,31096786,80.0,1,Quốc Lộ 1,trunk
3,3,2020-10-18T13:26:17.444Z,2020-10-18T13:26:17.444Z,5738158918,5738158912,66,31096786,80.0,1,Quốc Lộ 1,trunk
4,4,2020-10-18T13:26:17.452Z,2020-10-18T13:26:17.452Z,5738158912,5758104203,127,31096786,80.0,1,Quốc Lộ 1,trunk


In [76]:
new_train_df.head(1)

,_id,segment_id,date,weekday,period,LOS,s_node_id,e_node_id,length,street_id,max_velocity,street_level,street_name,segment_type,long_snode,lat_snode,long_enode,lat_enode
0,0,26,2021-04-16,4,period_0_30,A,366428456,366416066,116,32575820,40.0,4,Nguyễn Văn Bá,tertiary,106.768732,10.841506,106.769254,10.842422


In [77]:
filtered_train_df = new_train_df.drop(columns=[
    "s_node_id", "e_node_id", "length",
    "street_id", "max_velocity", "street_level", "street_name", "segment_type",
    "long_snode", "lat_snode", "long_enode", "lat_enode"
])
new_train_df.head()

,_id,segment_id,date,weekday,period,LOS,s_node_id,e_node_id,length,street_id,max_velocity,street_level,street_name,segment_type,long_snode,lat_snode,long_enode,lat_enode
0,0,26,2021-04-16,4,period_0_30,A,366428456,366416066,116,32575820,40.0,4,Nguyễn Văn Bá,tertiary,106.768732,10.841506,106.769254,10.842422
1,1,33,2020-08-02,6,period_23_30,C,366469460,3792257828,26,32575862,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808
2,2,33,2020-08-03,0,period_0_00,D,366469460,3792257828,26,32575862,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808
3,3,67,2021-03-09,1,period_9_30,B,366403668,5755066033,7,32575862,40.0,3,Đường số 5,secondary,106.768412,10.880817,106.768461,10.880771
4,4,67,2021-03-23,1,period_9_30,B,366403668,5755066033,7,32575862,40.0,3,Đường số 5,secondary,106.768412,10.880817,106.768461,10.880771


# Đổi lại đơn vị length trong segments và train, m -> km

In [78]:
new_segments_df["length"] = segments_df["length"] / 1000
segments_df.head()

,_id,created_at,updated_at,s_node_id,e_node_id,length,street_id,max_velocity,street_level,street_name,street_type
0,0,2020-10-18T13:26:17.365Z,2020-10-18T13:26:17.365Z,373543511,5468660805,114,31096786,80.0,1,Quốc Lộ 1,trunk
1,1,2020-10-18T13:26:17.400Z,2020-10-18T13:26:17.400Z,5468660805,5738158916,9,31096786,80.0,1,Quốc Lộ 1,trunk
2,2,2020-10-18T13:26:17.435Z,2020-10-18T13:26:17.435Z,5738158916,5738158918,23,31096786,80.0,1,Quốc Lộ 1,trunk
3,3,2020-10-18T13:26:17.444Z,2020-10-18T13:26:17.444Z,5738158918,5738158912,66,31096786,80.0,1,Quốc Lộ 1,trunk
4,4,2020-10-18T13:26:17.452Z,2020-10-18T13:26:17.452Z,5738158912,5758104203,127,31096786,80.0,1,Quốc Lộ 1,trunk


In [79]:
new_train_df["length"] = new_train_df["length"] / 1000
new_train_df.head()

,_id,segment_id,date,weekday,period,LOS,s_node_id,e_node_id,length,street_id,max_velocity,street_level,street_name,segment_type,long_snode,lat_snode,long_enode,lat_enode
0,0,26,2021-04-16,4,period_0_30,A,366428456,366416066,0.116,32575820,40.0,4,Nguyễn Văn Bá,tertiary,106.768732,10.841506,106.769254,10.842422
1,1,33,2020-08-02,6,period_23_30,C,366469460,3792257828,0.026,32575862,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808
2,2,33,2020-08-03,0,period_0_00,D,366469460,3792257828,0.026,32575862,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808
3,3,67,2021-03-09,1,period_9_30,B,366403668,5755066033,0.007,32575862,40.0,3,Đường số 5,secondary,106.768412,10.880817,106.768461,10.880771
4,4,67,2021-03-23,1,period_9_30,B,366403668,5755066033,0.007,32575862,40.0,3,Đường số 5,secondary,106.768412,10.880817,106.768461,10.880771


# Xây dựng tập dữ liệu tổng hợp dựa trên train

## Gộp street_type do bị tách ra street_type và segment_type

In [80]:
temp_street_df = new_streets_df.rename(columns={
    "_id": "street_id",
    "name": "street_name",
    "level": "street_level"
})

In [81]:
all_df = new_train_df.merge(
    temp_street_df,
    on=["street_id", "street_name", "street_level", "max_velocity"],
    how="inner"
)
all_df = all_df.rename(columns={
    "type": "street_type"
})
print(all_df.shape)
all_df.head()

(33441, 19)


,_id,segment_id,date,weekday,period,LOS,s_node_id,e_node_id,length,street_id,max_velocity,street_level,street_name,segment_type,long_snode,lat_snode,long_enode,lat_enode,street_type
0,0,26,2021-04-16,4,period_0_30,A,366428456,366416066,0.116,32575820,40.0,4,Nguyễn Văn Bá,tertiary,106.768732,10.841506,106.769254,10.842422,tertiary
1,1,33,2020-08-02,6,period_23_30,C,366469460,3792257828,0.026,32575862,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808,secondary
2,2,33,2020-08-03,0,period_0_00,D,366469460,3792257828,0.026,32575862,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808,secondary
3,3,67,2021-03-09,1,period_9_30,B,366403668,5755066033,0.007,32575862,40.0,3,Đường số 5,secondary,106.768412,10.880817,106.768461,10.880771,secondary
4,4,67,2021-03-23,1,period_9_30,B,366403668,5755066033,0.007,32575862,40.0,3,Đường số 5,secondary,106.768412,10.880817,106.768461,10.880771,secondary


## Gộp train với status

In [82]:
new_status_df = status_df.copy()

In [83]:
new_status_df["updated_at"] = pd.to_datetime(new_status_df["updated_at"])

In [84]:
new_status_df["time_bucket"] = new_status_df["updated_at"].dt.floor("30min")
new_status_df["period"] = (
    "period_" +
    new_status_df["time_bucket"].dt.hour.astype(str) + "_" +
    new_status_df["time_bucket"].dt.strftime("%M")
)
new_status_df["date"] = new_status_df["updated_at"].dt.strftime("%Y-%m-%d")
new_status_df = new_status_df.rename(columns={
    "_id": "status_id"
})
new_status_df.head()

,status_id,updated_at,segment_id,velocity,time_bucket,period,date
0,0,2020-07-03 14:55:31.869000+00:00,24845,20,2020-07-03 14:30:00+00:00,period_14_30,2020-07-03
1,1,2020-07-03 15:02:56.048000+00:00,33923,10,2020-07-03 15:00:00+00:00,period_15_00,2020-07-03
2,2,2020-07-04 08:15:52.696000+00:00,33824,5,2020-07-04 08:00:00+00:00,period_8_00,2020-07-04
3,3,2020-07-04 08:15:59.903000+00:00,33824,5,2020-07-04 08:00:00+00:00,period_8_00,2020-07-04
4,4,2020-07-04 08:16:08.201000+00:00,33824,5,2020-07-04 08:00:00+00:00,period_8_00,2020-07-04


In [85]:
all_df = all_df.merge(
    new_status_df,
    on=["segment_id", "date", "period"],
    how="inner"
)
all_df = all_df.rename(columns={
    "type": "street_type"
})
print(all_df.shape)
all_df.head()

(90938, 23)


,_id,segment_id,date,weekday,period,LOS,s_node_id,e_node_id,length,street_id,...,segment_type,long_snode,lat_snode,long_enode,lat_enode,street_type,status_id,updated_at,velocity,time_bucket
0,0,26,2021-04-16,4,period_0_30,A,366428456,366416066,0.116,32575820,...,tertiary,106.768732,10.841506,106.769254,10.842422,tertiary,89252,2021-04-16 00:55:31.333000+00:00,117,2021-04-16 00:30:00+00:00
1,1,33,2020-08-02,6,period_23_30,C,366469460,3792257828,0.026,32575862,...,secondary,106.761957,10.878650,106.762143,10.878808,secondary,892,2020-08-02 23:56:28.019000+00:00,34,2020-08-02 23:30:00+00:00
2,1,33,2020-08-02,6,period_23_30,C,366469460,3792257828,0.026,32575862,...,secondary,106.761957,10.878650,106.762143,10.878808,secondary,4501,2020-08-02 23:59:28.137000+00:00,26,2020-08-02 23:30:00+00:00
3,2,33,2020-08-03,0,period_0_00,D,366469460,3792257828,0.026,32575862,...,secondary,106.761957,10.878650,106.762143,10.878808,secondary,8110,2020-08-03 00:02:28.070000+00:00,2,2020-08-03 00:00:00+00:00
4,2,33,2020-08-03,0,period_0_00,D,366469460,3792257828,0.026,32575862,...,secondary,106.761957,10.878650,106.762143,10.878808,secondary,11719,2020-08-03 00:05:28.204000+00:00,35,2020-08-03 00:00:00+00:00


In [86]:
# Bỏ các trong tập dữ liệu tổng hợp

In [87]:
all_df.columns

Index(['_id', 'segment_id', 'date', 'weekday', 'period', 'LOS', 's_node_id',
       'e_node_id', 'length', 'street_id', 'max_velocity', 'street_level',
       'street_name', 'segment_type', 'long_snode', 'lat_snode', 'long_enode',
       'lat_enode', 'street_type', 'status_id', 'updated_at', 'velocity',
       'time_bucket'],
      dtype='str')

In [88]:
all_df = all_df.drop(columns={
    "_id", "street_id", "status_id",
    "time_bucket", "time_bucket"
})
all_df.head()

,segment_id,date,weekday,period,LOS,s_node_id,e_node_id,length,max_velocity,street_level,street_name,segment_type,long_snode,lat_snode,long_enode,lat_enode,street_type,updated_at,velocity
0,26,2021-04-16,4,period_0_30,A,366428456,366416066,0.116,40.0,4,Nguyễn Văn Bá,tertiary,106.768732,10.841506,106.769254,10.842422,tertiary,2021-04-16 00:55:31.333000+00:00,117
1,33,2020-08-02,6,period_23_30,C,366469460,3792257828,0.026,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808,secondary,2020-08-02 23:56:28.019000+00:00,34
2,33,2020-08-02,6,period_23_30,C,366469460,3792257828,0.026,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808,secondary,2020-08-02 23:59:28.137000+00:00,26
3,33,2020-08-03,0,period_0_00,D,366469460,3792257828,0.026,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808,secondary,2020-08-03 00:02:28.070000+00:00,2
4,33,2020-08-03,0,period_0_00,D,366469460,3792257828,0.026,40.0,3,Đường số 5,secondary,106.761957,10.878650,106.762143,10.878808,secondary,2020-08-03 00:05:28.204000+00:00,35


# Lưu data

In [89]:
PREPROCESS_PATH = "../data/preprocess"

In [90]:
nodes_df.to_csv(f"{PREPROCESS_PATH}/nodes.csv", index=False)

In [91]:
new_segments_df.to_csv(f"{PREPROCESS_PATH}/segments.csv", index=False)

In [92]:
new_streets_df.to_csv(f"{PREPROCESS_PATH}/streets.csv", index=False)

In [93]:
new_status_df.to_csv(f"{PREPROCESS_PATH}/segment_status.csv", index=False)

In [94]:
filtered_train_df.to_csv(f"{PREPROCESS_PATH}/train.csv", index=False)

In [95]:
all_df.to_csv(f"{PREPROCESS_PATH}/all.csv", index=False)